### PR TITLE
[ADR] ADR-037: Peer Behaviour

### PR DESCRIPTION
**Pull request by brapse**
* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

### CODE REVIEW

File: docs/architecture/adr-037-peer-behaviour.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    * Reduce the coupling of reactors and the Switch and the network
+      stack
+    * The responsibility of managing peer behaviour can be migrated to
+      a single component instead of split between the switch and the
+      address book.
+
+### Negative
+
+    * The first iteration will simply wrap the Switch and introduce a
+      level of indirection.
```
Comment (Liamsi) [inline docs/architecture/adr-037-peer-behaviour.md]: You could could also argue that this is not negative as you propose to improve the status quo incrementally (which is a good thing IMHO).

File: docs/architecture/adr-037-peer-behaviour.md [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+`switch.StopPeerForError(...)` when a peer misbehaves and 
+`switch.MarkAsGood(...)` when a peer contributes in some meaningful way. 
+While the switch handles `StopPeerForError` internally, the `MarkAsGood` 
+method delegates to another component, `p2p.AddrBook`. This scheme of delegation 
+across Switch obscures the responsibility for handling peer behaviour
+and ties up the reactors in a larger dependency graph when testing.
+
+## Decision
+
+Introduce a PeerBehaviour interface and concrete implementations which
```
Comment (Liamsi) [inline docs/architecture/adr-037-peer-behaviour.md]: nit:
```suggestion
Introduce a `PeerBehaviour` interface and concrete implementations which
```

File: docs/architecture/adr-037-peer-behaviour.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+provide methods for reactors to signal peer behaviour without direct
+coupling `p2p.Switch`.  Introduce a ErrPeer to provide 
+concrete reasons for stopping peers.
+
+### Implementation Changes
+
+PeerBehaviour then becomes an interface for signaling peer errors as well
+as marking peers for `good`.
```
Comment (Liamsi) [inline docs/architecture/adr-037-peer-behaviour.md]:
```suggestion
as for marking peers as `good`.
```

File: docs/architecture/adr-037-peer-behaviour.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+### Negative
+
+    * The first iteration will simply wrap the Switch and introduce a
+      level of indirection.
+
+### Neutral
+
+## References
+
+* [#2067](https://github.com/tendermint/tendermint/issues/2067): P2P Refactor
```
Comment (Liamsi) [inline docs/architecture/adr-037-peer-behaviour.md]: Nit: you can use the footnote style for linking (prob. there is an actual name for this) these to increase readability in the text above and to re-use the link. 
E.g.
```
Mention [#2067] in the text like this and use it again somewhere else, 
e.g. here: [#2067] and link to it once at the end (or somewhere):

[#2067]: https://github.com/tendermint/tendermint/issues/2067
```

File: docs/architecture/adr-037-peer-behaviour.md (hunk 1 above)
Comment (xla) [reply docs/architecture/adr-037-peer-behaviour.md]: :+1:

File: docs/architecture/adr-037-peer-behaviour.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+func (spb StorePeerBehaviour) Errored(peer Peer, reason ErrPeer) {
+    if _, ok := spb.pe[peer]; !ok {
+        spb.pe[peer] = []ErrPeer{reason}
+    } else {
+        spb.pe[peer] = append(spb.pe[peer], reason)
+    }
+}
+
+func (mpb *StorePeerBehaviour) GetPeerErrors() PeerErrors {
```
Comment (xla) [inline docs/architecture/adr-037-peer-behaviour.md]: Are the `Get` methods used to inspect the peer topology from test code?

File: docs/architecture/adr-037-peer-behaviour.md (hunk 4 above)
Comment (brapse) [reply docs/architecture/adr-037-peer-behaviour.md]: Do you mean like this?
```suggestion
* [#2067]: https://github.com/tendermint/tendermint/issues/2067
```
or possibly use  [the markdown footnote](https://www.markdownguide.org/extended-syntax/#footnotes) to link the text to the footnote to the actual link?

File: docs/architecture/adr-037-peer-behaviour.md (hunk 5 above)
Comment (brapse) [reply docs/architecture/adr-037-peer-behaviour.md]: Yes that's correct.

File: docs/architecture/adr-037-peer-behaviour.md (hunk 4 above)
Comment (Liamsi) [reply docs/architecture/adr-037-peer-behaviour.md]: > or possibly use the markdown footnote to link the text to the footnote to the actual link? 

This one. I think it makes the document slightly easier to read and you can re-use the reference in the text.

Comment (brapse) [reply docs/architecture/adr-037-peer-behaviour.md]: Maybe I'm missing something, but Github markdown doesn't appear to support that particular `[^#]` features but we can accomplish something similar by linking to the Reference section with sub text. 

4d7

Comment (Liamsi) [reply docs/architecture/adr-037-peer-behaviour.md]: Haha, where did that awesome meme go! I really just meant to replace the 2 (or more) occurrences of links to issue with what is described in https://github.com/tendermint/tendermint/pull/3539#discussion_r273869696
(and for above reasons)
github supports this.
But NVM. The PR is really good as it is and it was just a tiny nit.

File: docs/architecture/adr-037-peer-behaviour.md [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+as for marking peers as `good`.
+
+XXX: It might be better to pass p2p.ID instead of the whole peer but as
+a first draft maintain the underlying implementation as much as
+possible.
+
+```go
+type PeerBehaviour interface {
+    Errored(peer Peer, reason ErrPeer)
+    MarkPeerAsGood(peer Peer)
```
Comment (melekes) [inline docs/architecture/adr-037-peer-behaviour.md]: The first func now tells what happened (errored), the second one - what to do (mark). Maybe we should rename `MarkPeerAsGood` to `BecomeGood`?

Comment (Liamsi) [reply docs/architecture/adr-037-peer-behaviour.md]: Hmm, consistent would be sth like `BehavedWell`, `WasUseful`, `Contributed`? I think the actual name stems from the fact that the switch carries around an addrBook and the method there is `MarkGood`. Sound like we could improve the naming here @melekes 👍 Let's keep this in mind and wait for the implementation though.

Comment (xla) [reply docs/architecture/adr-037-peer-behaviour.md]: Good point! Let's ammend the ADR to not have it fall through the cracks during implementation. What about:

``` go
type PeerBehaviour interface {
    Behaved(id ID, reason GoodBehaviourPeer)
    Errored(id ID, reason ErrorBehaviourPeer)
}
```

@brapse & @milosevic Thoughts on this?

Comment (brapse) [reply docs/architecture/adr-037-peer-behaviour.md]: I really like the `Behaved`, `Errored` dichotomy!

The change from `peer p2p.Peer` -> `id p2p.ID` also further decouples the reactor from the network concern heavy peer interface. 👍 

The introduction of `GoodBehaviourPeer` also seems to complete the symmetry 👍 

Happy to amend:
image

**Pull request merged**